internal/rpc/rpc.go

```diff
@@ -0,0 +1,170 @@
+	Blocks int
+	Logs   int
+	Traces int
+}
+
+type Client struct {
+	RPCClient          *gethRpc.Client
+	EthClient          *ethclient.Client
+	SupportsTraceBlock bool
+	IsWebsocket        bool
+	URL                string
+	ChainID            *big.Int
+	BlocksPerRequest   BlocksPerRequestConfig
+}
+
+func Initialize() (*Client, error) {
+	rpcUrl := config.Cfg.RPC.URL
+	if rpcUrl == "" {
+		return nil, fmt.Errorf("RPC_URL environment variable is not set")
+	}
+	log.Debug().Msg("Initializing RPC")
+	rpcClient, dialErr := gethRpc.Dial(rpcUrl)
+	if dialErr != nil {
+		return nil, dialErr
+	}
+
+	ethClient := ethclient.NewClient(rpcClient)
+
+	rpc := &Client{
+		RPCClient:        rpcClient,
+		EthClient:        ethClient,
+		URL:              rpcUrl,
+		IsWebsocket:      strings.HasPrefix(rpcUrl, "ws://") || strings.HasPrefix(rpcUrl, "wss://"),
+		BlocksPerRequest: GetBlockPerRequestConfig(),
+	}
+	checkErr := rpc.checkSupportedMethods()
+	if checkErr != nil {
+		return nil, checkErr
+	}
+
+	chainIdErr := rpc.setChainID()
+	if chainIdErr != nil {
+		return nil, chainIdErr
+	}
+	return rpc, nil
+}
+
+func (rpc *Client) Close() {
+	rpc.RPCClient.Close()
+	rpc.EthClient.Close()
+}
+
+func (rpc *Client) checkSupportedMethods() error {
+	var blockByNumberResult interface{}
+	err := rpc.RPCClient.Call(&blockByNumberResult, "eth_getBlockByNumber", "latest", true)
+	if err != nil {
+		return fmt.Errorf("eth_getBlockByNumber method not supported: %v", err)
+	}
+	log.Debug().Msg("eth_getBlockByNumber method supported")
+
+	var getLogsResult interface{}
+	logsErr := rpc.RPCClient.Call(&getLogsResult, "eth_getLogs", map[string]string{"fromBlock": "0x0", "toBlock": "0x0"})
+	if logsErr != nil {
+		return fmt.Errorf("eth_getLogs method not supported: %v", logsErr)
+	}
+	log.Debug().Msg("eth_getLogs method supported")
+
+	var traceBlockResult interface{}
+	if config.Cfg.RPC.Traces.Enabled {
+		if traceBlockErr := rpc.RPCClient.Call(&traceBlockResult, "trace_block", "latest"); traceBlockErr != nil {
+			log.Warn().Err(traceBlockErr).Msg("Optional method trace_block not supported")
+		}
+	}
+	rpc.SupportsTraceBlock = traceBlockResult != nil
+	log.Debug().Msgf("trace_block method supported: %v", rpc.SupportsTraceBlock)
+	return nil
+}
+
+func (rpc *Client) setChainID() error {
+	chainID, err := rpc.EthClient.ChainID(context.Background())
+	if err != nil {
+		return fmt.Errorf("failed to get chain ID: %v", err)
+	}
+	rpc.ChainID = chainID
+	return nil
+}
+
+func (rpc *Client) GetFullBlocks(blockNumbers []*big.Int) []GetFullBlockResult {
+	var wg sync.WaitGroup
+	var blocks []RPCFetchBatchResult[common.RawBlock]
+	var logs []RPCFetchBatchResult[common.RawLogs]
+	var traces []RPCFetchBatchResult[common.RawTraces]
+
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+		blocks = RPCFetchBatch[common.RawBlock](rpc, blockNumbers, "eth_getBlockByNumber", func(blockNum *big.Int) []interface{} {
+			return []interface{}{hexutil.EncodeBig(blockNum), true}
+		})
```

Comment:
nit: defining callbacks while passing them as param makes it hard to read

Reply:
I agree on the readability, separated the param functions